pyccel/ast/cudaext.py

```diff
@@ -0,0 +1,309 @@
+        if memory_location not in ('device', 'host', 'managed'):
+            raise ValueError("The direction of the copy should be to 'host' or 'device'")
+        
+        # verify the type of is_async
+        if not isinstance(is_async, (LiteralTrue, LiteralFalse, bool)):
+            raise TypeError('is_async must be boolean')
+        
+        self._arg             = arg
+        self._shape           = arg.shape
+        self._rank            = arg.rank
+        self._dtype           = arg.dtype
+        self._order           = arg.order
+        self._precision       = arg.precision
+        self._memory_location = memory_location
+        self._is_async        = is_async
+        super().__init__()
+    
+    @property
+    def arg(self):
+        return self._arg
+
+    @property
+    def memory_location(self):
+        return self._memory_location
+
+    @property
+    def is_async(self):
+        return self._is_async
+
+class CudaThreadIdx(CudaInternalVar):
+    __slots__ = ()
+    pass
+class CudaBlockDim(CudaInternalVar):
+    __slots__ = ()
+    pass
+class CudaBlockIdx(CudaInternalVar):
+    __slots__ = ()
+    pass
+class CudaGridDim(CudaInternalVar):
+    __slots__ = ()
+    pass
+
+class CudaGrid(PyccelAstNode)               :
+    """
+    CudaGrid locate Thread In the GPU architecture Using CudaThreadIdx, CudaBlockDim, CudaBlockIdx
+    To calculate the exact index of the thread automatically.
+
+    Parameters
+    ----------
+    dim : NativeInteger
+        Represent the dimension where we want to locate our thread.
+
+    """
+    __slots__ = ()
+    _attribute_nodes = ()
+    def __new__(cls, dim=0):
+        if not isinstance(dim, LiteralInteger):
+            raise TypeError("dimension need to be an integer")
+        if dim not in (0, 1, 2):
+            raise ValueError("dimension need to be 0, 1 or 2")
+        expr = [PyccelAdd(PyccelMul(CudaBlockIdx(d), CudaBlockDim(d)), CudaThreadIdx(d))\
+                for d in range(dim.python_value + 1)]
+        if dim == 0:
+            return expr[0]
+        return PythonTuple(*expr)
+
+
+
+cuda_funcs = {
+    'array'             : PyccelFunctionDef('array'             , CudaArray),
+    'copy'              : PyccelFunctionDef('copy'              , CudaCopy),
+    'synchronize'       : PyccelFunctionDef('synchronize'       , CudaSynchronize),
+    'threadIdx'         : PyccelFunctionDef('threadIdx'         , CudaThreadIdx),
+    'blockDim'          : PyccelFunctionDef('blockDim'          , CudaBlockDim),
+    'blockIdx'          : PyccelFunctionDef('blockIdx'          , CudaBlockIdx),
+    'gridDim'           : PyccelFunctionDef('gridDim'           , CudaGridDim),
+    'grid'              : PyccelFunctionDef('grid'              , CudaGrid)
+}
+
+cuda_Internal_Var = {
+    'CudaThreadIdx' : 'threadIdx',
+    'CudaBlockDim'  : 'blockDim',
+    'CudaBlockIdx'  : 'blockIdx',
+    'CudaGridDim'   : 'gridDim'
+}
+
+cuda_constants = {
+
+}
```

Comment:
Is this necessary?

Reply:
as i was passing it to the `variables` attr of the class `Module` i made it like that but not i am just passing `[]` so i have removed it.
[9f989ff](https://github.com/pyccel/pyccel-cuda/pull/12/commits/9f989ff8e53005632eef4f421609c3d8ed4db490)